### ISSUE TITLE
Folders: Option to show folders hierarchically

### ISSUE THREAD

**Issue by coxtor** [photoprism/photoprism#1536]
Some users may have a personal folder hierarchy on their hard drive to find photos quickly. 
It would be great to retain the personal folder structure in the folder view page  or at least give the user the option to have it displayed in that way. The side panel pane could also act as a tree view and navigation option to make navigating easy. 
The current implementation makes it impossible to easily find pictures in large collections with many folders as the folder structure is flattend. 

----------------------------------------------------------------------------------------------------------------------------------------------
**Acceptance Criteria:**
* Add an option to toggle between flat and hierarchical view within the folders section
* TBD: Should edits/shares/downloads be recursive?

**Comment by graciousgrey** [comment]
The folder section is meant to be without hierarchy. 
We already have the original section (see https://demo.photoprism.org/library/files) to display all files (not only photos) in hierarchy.
We could add an option to the originals section to only show photos, like in other sections.

EDIT: After evaluating this again, we will probably go for a toggle between flat and hierarchical view in the Folders section.

**Comment by coxtor** [comment]
Thanks, I did not realize that that option was there. Nevertheless not having that option appears to be counterintuitive in a section called folders. 
:)

**Comment by AliMickey** [comment]
> 
> 
> The folder section is meant to be without hierarchy.
> We already have the original section (see https://demo.photoprism.org/library/files) to display all files (not only photos) in hierarchy.
> We could add an option to the originals section to only show photos, like in other sections.

Only issue with this is that in originals we do not get any of these options. 
![image](https://user-images.githubusercontent.com/60691199/134111156-5eb7360c-830e-482f-9a5a-fabd9a3fa960.png)

Sorting folders alphabetically would make it better, (my photos are stored by year 2000,2001,2001 etc.).

Hence, my PR https://github.com/photoprism/photoprism/pull/1361 to have a default page, so that we can land directly onto an untouched directory.

**Comment by vbc4ejyg** [comment]
I also have the same issue where I need to get to a specific folder in a particular year quickly, and there is no tree available to get me right there.  

> The folder section is meant to be without hierarchy.
> We already have the original section (see https://demo.photoprism.org/library/files) to display all files (not only photos) in hierarchy.
The originals section edits the photo details when you click on it to view.  I also need to browse to a folder within a specific year and go full screen to view it along with other pics in the same folder, but with the full features to move to the next photo tapping left and right (slideshow).  So it's really close however it edits instead of slideshow view.


![Capture](https://user-images.githubusercontent.com/39440750/134578641-cd9522ad-70c0-446c-9ac4-5be67b070f49.JPG)

When I look in "Folders", it's a bit of a mess of every folder I have over all years regardless of tree level.  I do love the ability to search it though!

**Comment by major-mayer** [comment]
I would really love to see such a feature in Photoprism.
Hierarchical folders in combination with a custom homepage would make Photoprism the perfect replacement (or better addition) for viewing photos in Nextcloud, which is very slow for me.

**Comment by GlassedSilver** [comment]
It should also be toggled whether you want to even see the subfolders in the view area or in the left hierarchy list only.

Basically just copy how Apple Aperture did it, the grandmaster of DAMs. (you cannot change my mind about this :p)

**Comment by ulysse132** [comment]
> The folder section is meant to be without hierarchy. We already have the original section (see https://demo.photoprism.org/library/files) to display all files (not only photos) in hierarchy. We could add an option to the originals section to only show photos, like in other sections.
> 
> EDIT: After evaluating this again, we will probably go for a toggle between flat and hierarchical view in the Folders section.

I agree, this is what I want but this section is missing the most important feature : the gallery. I mean, when you click on a photo  in "Folders" directory, it opens the photo in large screen and you can go to the next one using arrow keys, but in "Originals" directory, it opens an editing screen, not the photo itself.

I don't care about details, maybe a button could be enough. I would really like to view my photos using Originals, maybe it exists a way ?

**Comment by graciousgrey** [comment]
@ulysse132 It's not possible yet to view folders hierarchically , it will be possible once this issue is implemented :)

**Comment by ulysse132** [comment]
I've just seen that on the roadmap but it's stil on idea stage. There's no development scheduled yet. How to upvote this feature to make it pass to upcoming ?

**Comment by graciousgrey** [comment]
Our [Project Roadmap](https://link.photoprism.app/roadmap) shows what tasks are in progress and what features will be implemented next. You may give ideas you like a thumbs-up, so we know what's most popular.

Be aware that we have a zero-bug policy and do our best to help users when they need support or have other questions. This comes at a price though, as we can't give exact release dates for new features. 

Since sustained funding is key to quickly releasing new features, we encourage all users to support our mission by [signing up as a sponsor](https://photoprism.app/membership) or purchasing a [commercial license](https://photoprism.app/teams).

**Comment by AliMickey** [comment]
My personally sponsored and developed pull request from like a year still not merged. Probs out of date and needs conflicts to be resolved now. Was such a simple one as well.

**Comment by graciousgrey** [comment]
We understand that it is frustrating to wait so long for pull requests to be merged.
Please note that we do not merge pull requests if they do not cover the entire logic of a feature and are well tested, as this causes a massive support overhead on our end. 
This leads to long waiting times for pull requests that need additional work from us.

**Comment by AliMickey** [comment]
`We can finish this if you're not in a hurry.` - lastzero (https://github.com/photoprism/photoprism/pull/1361)

Didn't expect longer than a year. I would have just finished the remaining logic myself if I knew. 

Doesn't matter, product is getting better every day, we'll get there eventually. :)

**Comment by Dalmat** [comment]
Note as well that a major difference between the folder view and the library/originals view is that the folder view is a real gallery view with ability to switch between pictures by swiping. This is not available in the originals view, which means I will hardly scroll pictures in the current originals view

**Comment by wechsler42** [comment]
Hi there,
I would also like to thank the volunteers here which work together with the devs and Theresa that this feature will come to Photoprism. For me, who sits on a hierarchical folder structure spanning over a decade it is something that I would really love to see in Photoprism folder view.

**Comment by starcraft66** [comment]
Hey, I've been tracking this issue for a long time and this would fix my biggest pain point with photoprism. I don't have the time or webdev skills in this particular tech stacks to contribute the feature but I'd like to sponsor its development, is there a way to start a bounty for closing this issue specifically?

**Comment by graciousgrey** [comment]
Completing the work on multi-user support and merging existing pull requests is our top priority at the moment. Unfortunately, we do not have additional capacity to work on other tasks. In order for us to change this and grow our team, getting [sustained support](https://photoprism.app/membership) is very important.

**Comment by AliMickey** [comment]
Current revenue from Github + Patreon seems to be at $5000 as of now. 

Whats the size of the team now?

**Comment by lastzero** [comment]
Thanks for doing the math for us. Revenue was basically 0 the first years we both worked full time on this and now we also have infrastructure to pay for. I'm not even sure if it's minimum wage yet.

**Comment by lastzero** [comment]
Let us know if you want to join our team full time for free! You're welcome!

**Comment by lastzero** [comment]
By the way, we also pay taxes on this. If you didn't know.

**Comment by AliMickey** [comment]
I didn't accuse you of mismanaging funds, nor did I intend to provoke. Was simply interested in the size of the team based off the revenue just from the two sources.

Also do I get part of the revenue if I join 😅.

**Comment by lastzero** [comment]
The team is self-funded because no one else was doing this work and we realized that Big Tech is mismanaging our personal data. Once we have revenue left over to spend on a team, you are welcome to apply for a position! Can I ask how much you currently make so we know if the position would be attractive to you?

**Comment by kirkanos** [comment]
This would be a very nice Feature. It's nearly the only point I miss in PhotoPrism. Nice work, I love it.

**Comment by JGKle** [comment]
Hi - just checking out PhotoPrism, as I'm looking for an open-source alternative to Synology Photos (specifically for accessing my library from an Android device).

Like others above, my library is organized via nested folder names. I saw the suggestion above to try https://demo.photoprism.org/library/files, but this just seems to redirect to https://demo.photoprism.org/library/albums, which doesn't seem to be hierarchical.  Perhaps I misunderstood? From what I could tell, it doesn't seem like PhotoPrism currently has a way to navigate, which directly reflects how the files are stored on disk. As that's core to my use case, I wanted to check just in case I misunderstood, before continuing my search elsewhere :)

Thanks in advance!

**Comment by graciousgrey** [comment]
@metal450 the link has changed it's https://demo.photoprism.app/library/index/files

It does not cover all use cases that's why this issue is on our roadmap :)

**Comment by JGKle** [comment]
Ah, gotcha - thanks for the quick reply.

**Comment by kkovaletp** [comment]
I'm just curious whether there are some updates on this topic since the last ~6 months from the previous message...
For me this is the top feature, I'm waiting for: I use Photoprism for several months and so far this is the only serious gap in functionality, I see - I still use Samba file share in some of my use cases because of that.

**Comment by graciousgrey** [comment]
We haven't had any resources left to work on it yet. 

Our [Project Roadmap](https://link.photoprism.app/roadmap) shows what tasks are in progress and what features will be implemented next. You may give ideas you like a thumbs-up, so we know what's most popular.

Be aware that we have a zero-bug policy and do our best to help users when they need support or have other questions. This comes at a price though, as we can't give exact release dates for new features.

We kindly ask you not to use Github Issues for general questions or personal opinions, as all contributors will receive an email each time. You are welcome to use Github Discussions for that.

Thank you very much ❤️

**Comment by UnasZole** [comment]
Hello,

I too have a large collection of photos organised in hierarchical folders, and I'm considering using photoprism to present them.
Reading the documentation, photoprism seems to have the cleanest way to manage the photos on the backend side (with separate "originals" and "sidecar" folder, ability to convert raws with rawtherapee using given pp3 files, etc.) among other gallery software I've checked.
However, the ability to browse all my photos with a hierarchical folder view is essential to me.

A view similar to what you have in Library > Originals could do the trick, but as @ulysse132 mentioned when clicking on actual photos there you end up on a metadata edition screen, not a photo viewer.
Also, I need a clarification here : the tab is called "Originals", so does it mean that it browses directly the Originals folder ? Meaning that it would ignore all the sidecar files generated, since they are in distinct folders ?
I have a number of RAWs in my gallery, and the generation of sidecar JPG files from them is a key feature. So it's important for me that this "hierarchical folder view" would still display the generated sidecars for each photo, and not just the original files.

**Comment by platima** [comment]
@UnasZole great questions!

@graciousgrey I am not sure what your funding is like, but can we sponsor this at all? Eg happy to chuck in $250 AUD to just make this happen. I know it's not much, but I figure it's ~5-10 hours of development time.

Else if anyone else has done it, but PRs not approved, I'm happy to test it.

This is the closest solution we've found for our needs, and pretty much the only feature missing! Great work and thank you kindly <3

**Comment by mcmara** [comment]
The ability to browse my large collection of photos with a hierarchical folder view is essential to me as well. PhotoPrism seems the best photo manager out there, although it is missing this fundamental feature. A flat view of thousands of folders is simply too dispersive.

**Comment by bmondream** [comment]
@mcmara Isn't the `/library/index/files` endpoint exactly that? You can get there from the side panel by clicking `Library` -> `Originals`. I just discovered it as I was reading through this issue, and ~~that's exactly what I was looking for.~~

I'm impressed by the quality of this project and I'm very grateful for it 🙂

edit: just realized you can't navigate between photos inside a directory, and you're shown metadata upon clicking on a photo instead of the whole photo. I guess it's not what I had in mind

**Comment by vbc4ejyg** [comment]
Correct.  I was just logging in to stay that it's NOT what you're/we're looking for. 😂

**Comment by Kansai53** [comment]
Just to chime in since I see some activity here - I've also been following this for more than 2 years. Unfortunately still haven't been able to use photoprism as this is a really a fundamental need based on how my photos are organized. Hope it gets some attention soon :)

**Comment by wechsler42** [comment]
Well, I am also waiting since more than 1 year on this at least for me very important feature. I just had a look in the roadmap section.  Folders: Option to show folders hierarchically #1536  is still there but it resides still in the "Ideas" column, see https://github.com/orgs/photoprism/projects/5/views/1?filterQuery=1536

**Comment by lukasradek** [comment]
I sort of kind of maybe hoped that this will be implemented after 4 years 😀. This is cruatial for people who categorize their photos using directories.

**Comment by sic6SaNdMaN** [comment]
> I sort of kind of maybe hoped that this will be implemented after 4 years 😀. This is cruatial for people who categorize their photos using directories.

You mean what every sane person does?
Yeah...
Can't figure out, why this is not the first thought when a project like this gets startet...

Like "OK, let's make a photo app... when we start, organizing photos, let's first have a look how they are stored on disc at the moment at most peeps PCs".

But I think everything is focussed on phones nowadays...

Don't get me wrong... I'm glad for every free/opensource software and appreciate the time and effort that is put into those projects.
But I simply can't get my head around, how this could be overseen or decided not to be important in the first place.

**Comment by vbc4ejyg** [comment]
It is true... almost everything is focused on phones these days. And most people don't even have a computer now.  Likely because they had to pay $1,500 for their iPhone every 2 or 3 years and there just isn't room to spend more...

For those of us around before metadata even existed, and some pictures are scanned in, folders are the only way. 1/5 of my photos have metadata, the rest do not, and it's an absolute mess in most apps.  So yeah, I am still hoping and patiently waiting for folders as well.  I like that it's at least an "idea" now, but I'm not counting on it.  I really appreciate what has been done on this app thus far and really hope that I can officially support it in the future when it meets my needs.  But one thing I know for sure is that I can't spend the remaining years of my life going back and editing in metadata dates, times, locations/GPS coordinates to make it all work.  It would take me 10 years! 

Photoprism seems to be in pretty good shape now, so I hope there can be some efforts in the near future to make it more "backwards compatible" for non-metadata assets.

**Comment by lukasradek** [comment]
I mean... you don't even have to be metadata agnostic to be in need of folders. If you want to group your photos together in any sort of a portable way (meaning not tied to a single ecosystem) - you need folders. Or metadata tags of course, but folders are much more usable.

**Comment by sic6SaNdMaN** [comment]
This.

I tried so many apps recently...
PhotoPrism
Immich
Piwigo
Pigallery2
and many more, can't even remember all of them.
Had to cleanup my Pi and the database afterwards for hours to get rid of all the data scrap flying around...

In the end, that doesn't even matter ;), because I ended up using Nextcloud.
Mounting my Photos-folder into the Photos-Folder from Nextcloud, updating the file index and let it generate thumbnails in advance for fast browsing.
Is it perfect? No.
Is it working? Mostly.
Was it a big hazzle and did I have to change my existing photos and hierarchical folder structure? No.

So... I'll stick with that for now...

**Comment by Parmenide78** [comment]
On nextcloud you have the "Memories" App too ^^

**Comment by sic6SaNdMaN** [comment]
True, you do.
But you also don't have to use it.
You can even uninstall it completely.

**Comment by JGKle** [comment]
Fwiw after 2.5yrs I gave up waiting, & am now using Immich - which *does* actually support folder view (both in their web UI and their mobile app). See: https://docs.immich.app/features/libraries. Depending on your usecase it works great (I just wanted a way to access my whole photo collection from mobile, and be able to navigate in a sane manner - aka as it's actually organized on disk, not based on partial/incomplete metadata).

**Comment by mcmara** [comment]
I also kept Photoprism, Immich and Nextcloud at the same time for evaluation, with my large collection dating back to the year 1985, when I was using a SLR camera with slides. While Photoprism ignored this fundamental capability, Immich has introduced a very basic folders view about a year ago. Few months ago the folder view was improved eliminating the problem of name truncation, so now it seems doing its job sufficiently well. For this reason I abandoned Photoprism, at least until I see this issue resolved. Nextcloud memories is OK, but the scope of Nextcloud is much larger than media handling and for that reason is not my first choice.

**Comment by sic6SaNdMaN** [comment]
Hmmm... tested Immich just a few weeks/month ago and decided, that it also didn't fit my needs - don't know exactly anymore why it didn't, though...
Perhaps I'll read the changelogs and release notes and give it another try.

**Comment by vbc4ejyg** [comment]
Thanks for the Immich tip and I will check it out!! 

I do hope Photoprism can do this someday, as it's a really great app and there active development which I love.  Kudos to them for keeping it going.

**Comment by platima** [comment]
> Hmmm... tested Immich just a few weeks/month ago and decided, that it also didn't fit my needs - don't know exactly anymore why it didn't, though... Perhaps I'll read the changelogs and release notes and give it another try.

Yeah I was much the same. I tested it maybe a year ago, and was not hugely happy, but when I tested it again recently - and took more time to read the docs and set it up properly - I was mildly surprised at how well it works!

GL, and happy holidays everyone.

PS: Michael and Theresa, your product is amazing - do not get me wrong. I sponsored the project in July 2024 to support it, and I specifically requested this feature then. The problem really appears to be the community engagement, and the prioritisation. At the time I even offered a few hundred dollars to motivate things, but it did not move this request at all. I understand you have your roadmap/timeline, and other bugs and ideas to tend to, but I think when it's #4 on the list it can be really worth paying more attention to. Not to mention the ones above it. Best of luck either way!

**Comment by jdevora** [comment]
I installed it and was almost decided, but I couldn’t figure out this basic feature…
After looking around, it seems to be a “religious” thing — the lead developer had a bad experience with tree-like structures in the past, and I think it will take a few years to get implemented. :-(

Very nice product, but I’ve spent countless hours organizing my 30+ years photo library, and I’m not willing to do it again.

**Comment by graciousgrey** [comment]
Thank you for your feedback! 

We know that a hierarchical folder view in addition to the existing folders view is important to many users and we plan to implement it. The reason it is not implemented yet is that we have worked on other important and highly requested features such as batch editing and improved facial recognition among others.
